src/psyclone/domain/nemo/transformations/create_nemo_loop_trans.py

```diff
@@ -0,0 +1,148 @@
+            Literal[value:'1', Scalar<INTEGER, UNDEFINED>]
+            Literal[value:'10', Scalar<INTEGER, UNDEFINED>]
+            Literal[value:'1', Scalar<INTEGER, UNDEFINED>]
+            Schedule[]
+                0: Assignment[]
+                    ArrayReference[name:'tmp']
+                        Reference[name:'ji']
+                    BinaryOperation[operator:'MUL']
+                        Literal[value:'2', Scalar<INTEGER, UNDEFINED>]
+                        Reference[name:'ji']
+
+    As shown above, the resulting Schedule now contains a NemoLoop, indicated
+    by the "type='lon'" (for 'longitude') annotation for the Loop node.
+
+    """
+    @property
+    def name(self):
+        '''
+        :returns: the name of the transformation.
+        :rtype: str
+
+        '''
+        return type(self).__name__
+
+    def validate(self, node, options=None):
+        '''
+        Check that the supplied node is a valid target for this transformation.
+
+        :param node: the target of the transformation.
+        :type node: :py:class:`psyclone.psyir.nodes.Node`
+        :param options: a dictionary with options for \
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dict of string:values or None
+
+        :raises TransformationError: if the supplied node is not a Routine.
+
+        '''
+        super(CreateNemoLoopTrans, self).validate(node, options=options)
+
+        if not isinstance(node, Loop):
+            raise TransformationError(
+                "Error in CreateNemoLoopTrans transformation. The supplied "
+                "node should be a PSyIR Loop but found '{0}'".format(
+                    type(node).__name__))
+
+    def apply(self, loop, options=None):
+        '''
+        Takes a generic PSyIR Loop node and replaces it with a NemoLoop.
+        The original loop object is deleted.
+
+        :param sched: the Loop node to be transformed.
+        :type sched: :py:class:`psyclone.psyir.nodes.Loop`
+        :param options: a dictionary with options for \
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dict of string:values or None
+
+        :returns: the new NemoLoop object.
+        :rtype: :py:class:`psyclone.nemo.NemoLoop`
+
+        '''
+        self.validate(loop, options=options)
+
+        # Convert a generic loop into a NEMO Loop by creating a new
+        # NemoLoop object and inserting it into the PSyIR.
+        nodes = loop.pop_all_children()
+        new_loop = NemoLoop.create(loop.variable,
+                                   nodes[0], nodes[1], nodes[2],
+                                   nodes[3].pop_all_children())
+        loop.replace_with(new_loop)
+        del loop
```

Comment:
I have the same question here about if the `del` is necessary and what is the expectation if there is an already existing reference to the original node?

Reply:
Yes (or no), removed.